src/coins.h

```diff
@@ -230,6 +230,34 @@ class CCoinsViewCursor
     uint256 hashBlock;
 };
 
+struct CoinsViewCacheCursor {
+public:
+    CoinsViewCacheCursor(size_t& usage, CoinsCachePair& sentinel, CCoinsMap& map) noexcept
+        : m_usage(usage), m_sentinel(sentinel), m_map(map) {}
+
+    inline CoinsCachePair* Begin() const noexcept { return m_sentinel.second.Next(); }
+
+    inline CoinsCachePair* End() const noexcept { return &m_sentinel; }
+
+    inline CoinsCachePair* Next(CoinsCachePair& current, bool erase) noexcept
+    {
+        const auto next_entry{current.second.Next()};
+        if (!erase) {
+            if (current.second.coin.IsSpent()) {
+                m_usage -= current.second.coin.DynamicMemoryUsage();
+                m_map.erase(current.first);
```

Comment:
isn't the parameter naming off if `not erase`, will lead to `map.erase`?

Reply:
Ahh yes (this comment)[https://github.com/bitcoin/bitcoin/pull/28280/files#diff-f0ed73d62dae6ca28ebd3045e5fc0d5d02eaaacadb4c2a292985a3fbd7e1c77cL269-L270] is removed in this patch, and should actually probably go right here. Perhaps I do this in a follow-up to not invalidate your ACK?

Reply:
> variation of it should actually probably go right here

Maybe, but I would still be surprised, regardless of the context, if `not X` leads to `X`. Could the parameter explain it better than "erase", so that the reviewer isn't surprised that it means "keep"?

> invalidate your ACK

I expect that I will have to review and ack your changes many more times before it's merged, don't worry about it!

Reply:
> I expect that I will have to review and ack your changes many more times before it's merged

I certainly hope that isn't the case :sweat_smile:

Reply:
@paplorinc what do you think a better name for this parameter would be?

Reply:
In https://github.com/bitcoin/bitcoin/blob/dddb59a63d90a108e7acefacdb6689a07324a6d7/src/coins.cpp#L193, we're calling `Next` with `erase`, which simply return `current.second.Next()`  without any mutation, because of https://github.com/bitcoin/bitcoin/blob/dddb59a63d90a108e7acefacdb6689a07324a6d7/src/coins.h#L245, right?

This is the part I don't understand, especially since later in https://github.com/bitcoin/bitcoin/blob/dddb59a63d90a108e7acefacdb6689a07324a6d7/src/coins.cpp#L207, we're moving when `erase = true` because we've already erased it - right?

So why is `erase == true || it->second.coin.IsSpent() == true` an erasure in one place, but `erase == false && current.second.coin.IsSpent() == true` an erasure in the other method? Shouldn't the Next method rather be something like this instead:

```C++
inline CoinsCachePair* Next(CoinsCachePair& current, bool erase) noexcept
{
    const auto next_entry{current.second.Next()};
    if (erase || current.second.coin.IsSpent()) {
        m_usage -= current.second.coin.DynamicMemoryUsage();
        m_map.erase(current.first);
    } else {
        current.second.ClearFlags();
    }
    return next_entry;
}
```
The related tests seem to be passing this way, which at least suggests a missing test case, if not an invalid erasure.

Reply:
Great question!

When `erase = true`, we will be erasing the entire map at the end of `BatchWrite`. When `erase = false`, we want to keep all unspent coins in the map and have their flags cleared.

While looping through `CCoinsMap::iterator`s, we can just do `it = erase ? mapCoins.erase(true) : std::next(it)`, and then deleting all spent coins and clear their flags right after we return from `BatchWrite` when `erase = false`.

Now that we are looping through `CoinsCachePair`s though, we can't call `mapCoins.erase(it)` because a `CoinsCachePair` can not be coerced back to a `CCoinsMap::iterator`. So, instead of erasing while we iterate when `erase = true`, we just keep all entries and call `mapCoins.clear()` after we return from `BatchWrite`.
But, as a performance optimization for the case where `erase = false`, we can also delete the spent coins in the same loop so we don't have to loop through them again when we return. So in this case we can erase by the map key instead of the map iterator. This is slower because we have to hash the key and lookup the entry in the map, whereas with the iterator we could just delete it directly.

When `erase = true` we *don't* want to erase by the key, because it is slower than calling `clear` at the end of the loop.

Reply:
>  , we're moving when erase = true because we've already erased it - right?

We're moving because we are going to erase it in either case, not that we have yet. We can't move an erased entry because it will have already been destroyed. And we can't use a moved entry later because it is invalidated by the move. So when we want to keep the entry, we need to copy it and not move out of it.
When `erase = true`, we will be erasing the entry after we return, so all entries can be moved. When `coin.IsSpent()`, we will be erasing it when `erase = false` because we only want to keep unspent entries.

I did have a commit before that changed `erase` to `will_erase`, I can't remember why I removed it. Would that make it clearer if I renamed that again?

Reply:
I will have to read that again tomorrow, it's the end of the day here :/
`will_erase` could _help_, but I find it hard to keep track of so many independent mutations, I'm hoping there's a simpler way to do this...
Do you think you could add a test that at least doesn't pass for my suggested code above? Maybe that will help me understand why something this simple (iteration with/without simple/bulk cleanup) has this many loose parts, where the exact same parameter means the opposite from one line to the next...

Reply:
You're right, the logic for moving or not should not leak into BatchWrite. Perhaps we can get rid of `erase` in BatchWrite and have the cursor return a pair - the iterator and a bool for whether it can move out or not?

Reply:
sounds better, would `Next` become a `const` in that case?

Reply:
Err, no `Next` would still conditionally erase entries. This would remove the `erase || it->second.coin.IsSpent()` condition in `BatchWrite`.

Reply:
Could we do that and delegate the deletion to a separate "service"? Seems complicated enough to warrant better encapsulation

Reply:
I had it like that before but others suggested to move everything into 1 loop instead of looping through the dirty entries again after BatchWrite to clear flags and delete spent entries.

Reply:
Can we have it both, single loop, but the mutation and its side-effects (like the move) separated clearly in the code from everything else?

Reply:
Will give it some thought

Reply:
Have you given it any more thought? I've tested it since, so i'll ack it, and will reack when you need more changes.

Reply:
I haven't been able to come up with something that I think would be an improvement. I did split the second last commit up into two commits though. Maybe that will make it more clear when reviewing.

Reply:
While reviewing (and before reading this discussion) I also got stuck on this. Because of the expectations the naming raises, it seems like a potential footgun (future code may try to iterate over the linked list without intending to erase entries).
I think the side effect should be documented more clearly for the `Next()` function, maybe even rename to `NextAndMaybeErase()` or something similar.

Reply:
This comment thread is pointing to an out of date version of the PR. There's more documentation on this struct and methods now, as well as changing a variable and field name. Maybe it is more clear now?

Reply:
Renamed the method to be `NextAndMaybeErase`.

Reply:
Also added more comments around this struct.